File: lab_02/PCY.py
import sys
from math import floor
from itertools import combinations


def first_pass():
    baskets = []
    item_count = {}

    for line in sys.stdin:
        items = [int(x) for x in line.strip().split(' ')]
        baskets.append(items)

        for item in items:
            item_count[item] = item_count.get(item, 0) + 1

    return baskets, item_count, len(item_count)


def second_pass(baskets, item_count, item_count_len, prag, b):
    pretinci = {}
    for basket in baskets:
        for i, j in combinations(basket, r=2):
            if item_count[i] >= prag and item_count[j] >= prag:
                k = (i * item_count_len + j) % b
                pretinci[k] = pretinci.get(k, 0) + 1

    return pretinci


def third_pass(baskets, item_count, item_count_len, prag, b, pretinci):
    parovi = {}
    for basket in baskets:
        for i, j in combinations(basket, r=2):
            if item_count[i] >= prag and item_count[j] >= prag:
                k = (i * item_count_len + j) % b
                if pretinci[k] >= prag:
                    parovi[(i, j)] = parovi.get((i, j), 0) + 1

    return parovi


def main():
    n = int(sys.stdin.readline())  # broj kosara
    s = float(sys.stdin.readline())  # prag podrske
    b = int(sys.stdin.readline())  # broj pretinaca
    prag = floor(s * n)

    baskets, item_count, item_count_len = first_pass()
    pretinci = second_pass(baskets, item_count, item_count_len, prag, b)
    parovi = third_pass(baskets, item_count, item_count_len, prag, b, pretinci)

    print(len(pretinci))
    print(len(parovi))
    [print(parovi[i]) for i in sorted(parovi, key=parovi.get, reverse=True)]


if __name__ == '__main__':
    main()

File: lab_02/java/src/PCY.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class PCY {

    public static void main(String[] args) throws IOException {
        int b, threshold;

        final List<List<Integer>> baskets = new ArrayList<>();
        final Map<Integer, Integer> itemCount = new HashMap<>();
        final Map<Integer, Integer> compartments = new HashMap<>();
        final Map<List<Integer>, Integer> pairs = new HashMap<>();

        try (final BufferedReader reader = new BufferedReader(new InputStreamReader(System.in))) {
            int n = Integer.parseInt(reader.readLine());
            double s = Double.parseDouble(reader.readLine());
            b = Integer.parseInt(reader.readLine());
            threshold = (int) Math.floor(s * n);

            for (int i = 0; i < n; i++) {
                final List<Integer> basket = Arrays.stream(reader.readLine().trim().split(" "))
                    .map(Integer::parseInt)
                    .collect(Collectors.toList());

                baskets.add(basket);
                basket.forEach(item -> itemCount.compute(item, (key, val) -> val == null ? 1 : val + 1));
            }
        }

        final int itemCountSize = itemCount.size();

        for (final List<Integer> basket : baskets) {
            final int basketSize = basket.size();
            for (int i = 0; i < basketSize; i++) {
                for (int j = i + 1; j < basketSize; j++) {
                    final int item1 = basket.get(i);
                    final int item2 = basket.get(j);

                    if (itemCount.get(item1) >= threshold && itemCount.get(item2) >= threshold) {
                        final int k = (item1 * itemCountSize + item2) % b;
                        compartments.compute(k, (key, val) -> val == null ? 1 : val + 1);
                    }
                }
            }
        }

        for (final List<Integer> basket : baskets) {
            final int basketSize = basket.size();
            for (int i = 0; i < basketSize; i++) {
                for (int j = i + 1; j < basketSize; j++) {
                    final int item1 = basket.get(i);
                    final int item2 = basket.get(j);

                    if (itemCount.get(item1) >= threshold && itemCount.get(item2) >= threshold) {
                        final int k = (item1 * itemCountSize + item2) % b;
                        if (compartments.getOrDefault(k, 0) >= threshold) {
                            pairs.compute(List.of(item1, item2), (key, val) -> val == null ? 1 : val + 1);
                        }
                    }
                }
            }
        }

        System.out.println(itemCountSize * (itemCountSize - 1) / 2);
        System.out.println(pairs.size());
        pairs.values().stream().sorted(Comparator.reverseOrder()).forEach(System.out::println);
    }

}
